streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -374,21 +374,21 @@ private NamedCacheMetrics(final StreamsMetrics metrics, final String name) {
 
             // add parent
             final Sensor parent = this.metrics.registry().sensor(opName, Sensor.RecordingLevel.DEBUG);
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(parent, this.metrics.registry().metricName(opName + "-avg", groupName,
-                "The average cache hit ratio.", allMetricTags), new Avg());
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(parent, this.metrics.registry().metricName(opName + "-min", groupName,
-                "The minimum cache hit ratio.", allMetricTags), new Min());
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(parent, this.metrics.registry().metricName(opName + "-max", groupName,
-                "The maximum cache hit ratio.", allMetricTags), new Max());
+            parent.add(this.metrics.registry().metricName(opName + "-avg", groupName,
+                    "The average cache hit ratio.", allMetricTags), new Avg());
+            parent.add(this.metrics.registry().metricName(opName + "-min", groupName,
+                    "The minimum cache hit ratio.", allMetricTags), new Min());
+            parent.add(this.metrics.registry().metricName(opName + "-max", groupName,
+                    "The maximum cache hit ratio.", allMetricTags), new Max());
 
             // add child
             hitRatioSensor = this.metrics.registry().sensor(opName, Sensor.RecordingLevel.DEBUG, parent);
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(hitRatioSensor, this.metrics.registry().metricName(opName + "-avg", groupName,
-                "The average cache hit ratio.", metricTags), new Avg());
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(hitRatioSensor, this.metrics.registry().metricName(opName + "-min", groupName,
-                "The minimum cache hit ratio.", metricTags), new Min());
-            ((StreamsMetricsImpl) metrics).maybeAddMetric(hitRatioSensor, this.metrics.registry().metricName(opName + "-max", groupName,
-                "The maximum cache hit ratio.", metricTags), new Max());
+            hitRatioSensor.add(this.metrics.registry().metricName(opName + "-avg", groupName,
+                    "The average cache hit ratio.", metricTags), new Avg());
+            hitRatioSensor.add(this.metrics.registry().metricName(opName + "-min", groupName,
+                    "The minimum cache hit ratio.", metricTags), new Min());
+            hitRatioSensor.add(this.metrics.registry().metricName(opName + "-max", groupName,
+                    "The maximum cache hit ratio.", metricTags), new Max());
```

Comment:
These metrics never get removed. Is that ok?

Reply:
Ok, when I took another look, I found that `hitRatioSensor` does get removed, but its parent doesn't. Also neither sensors have scoped names, so every `record()` will actually update *all* hit ratio metrics for *all* caches.

That seems like a bigger deal, so I've already prepared a follow-up PR. I'll send it next once this one is merged.